### ISSUE TITLE
Rename `Context`?

### ISSUE THREAD

**Issue by jnicklas** [thefrontside/effection#207]
One design choice that I made in mini-effection was renaming `Context` to `Task`. This aligns with [Rust's async/await](https://doc.rust-lang.org/std/task/index.html), with the [CLR `task` class](https://docs.microsoft.com/en-us/cpp/parallel/concrt/reference/task-class?view=msvc-160), and [with ember-concurrency](http://ember-concurrency.com/docs/tutorial/refactor) (somewhat). 

I personally don't find `Context` to be a very intuitive name for the concept that it represents.

Maybe contrasting this with the usage of `Context` in GraphQL makes sense, here a context object is an object that represents some type of environment, a bundle of data that the query can access. This feels in line with what I would expect a context to be. But currently a context represents a sequence of asynchronous operations (e.g. it can be halted).

In other words, I would expect `Context` to define the set of memory/storage associated with a sequence of operations, not the sequence itself.

I think `ExecutionContext` is somewhat better, but it's both wordy and awkward, and still suffers a little bit from the same issues. `Task` is short and snappy, yet fits well for the intended purpose.

**Comment by dagda1** [comment]
@jnicklas I first came across `async` `await` in .NET and I actually thought `Context` in effection was an environment and not a .NET like `Task<T>` so I am definitely in favour of. the name change.

**Comment by taras** [comment]
Effection's Context might be closer to React's context in concept, but I find that proximity confusing. In general, I find the use of Context in programming to be confusing. It's like Data - it's so broad that anything qualifies which leads not knowing what it actually is.

@jnicklas can you use "task" in a sentence to show an example of what explaining Effection using task would be like?

**Comment by dagda1** [comment]
I actually was thinking an `Operation<T>` was analogous to a `Task<T>` in .NET land.

**Comment by jnicklas** [comment]
@dagda1 it's somewhere in between, I think. Since .Net async functions immediately start executing when invoked, what is returned from the async function is actually an object representing the currently running thread of execution. This is in contrast to e.g. Rust, where Future does nothing until it starts to be polled by a Task, which is more similar to what we are doing in Effection. I don't think .Net separates these two concepts very clearly, but i might help to think of our concept of `Operation` as akin to an async function type in .Net, whereas `Task`/`Context` is what such a function returns.

@taras maybe something like:

> A Task is responsible for running an operation to completion, as well as managing the lifecycle of the operation and of any spawned sub-tasks.

**Comment by jnicklas** [comment]
@taras it might help to take a look at the [`Task` class](https://github.com/jnicklas/mini-effection/blob/master/src/task.ts) in mini-effection, I think it is easier to understand its purpose since it does not mix in any of the responsibilities of the operation.

**Comment by taras** [comment]
It sounds like in mini-effection, it's tasks and subtasks as an alternative to operations and their context. Task/sub-task is more clear and simple.

**Comment by cowboyd** [comment]
I share the dislike for `Context`as being  too abstract and empty a term, but I think my biggest sticking point is that Task implies something running, which they often are, especially in a server side application, but at the same they can be a static tree of things that do little more than accumulate and manage state. 
 
The other unresolved issue for me is what do we call the higher order ember-concurrency supervisor-like thing that we build on top of Context-y/Task-y thing that manages a set of running operations in a structured way with attributes like `maxConcurrency`, `dropLatest`, `dropFirst` ? I think having an idea of what that would be named would also provide good context for naming this thing.
